### PR TITLE
feat: adds migration api and queries

### PR DESCRIPTION
**Pull request by jscyo**
## Summary of change
(A few sentences about this PR)

## Related issues
- Link to issue1 here
- Link to issue1 here

## Test Plan
(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes
(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates
- [ ] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
   - In `build.gradle`
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them in `implementationDependencies.json`.
- [ ] Issue this PR against the latest non released version branch.
   - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
   - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR
- [ ] Convert all flows to tests
- [ ] Add firebase Scrypt support
   - [ ] How do other providers to password hash migration from firebase
   - [ ] what prefix do we add to the password hash
   - [ ] Does firebase always give the same length password hash or is it configurable
   - [ ] add tests for verification 
- [ ] see if firebase allows you to change the key, rounds.. config. If they do, how to manage backward compatibility? If they don't allow you to change, have they changed it in the past and have they maintained backwards compatability
- [ ] What is the nature of firebase verification, is it similar to argon? Do we need to use pool size like with argon? If it is do we need to give a config to change the pool size.
- [ ] speak to Mihaly about Scrypt and encryption(Cost and edge case etc...).

### CODE REVIEW

File: src/main/java/io/supertokens/emailpassword/PasswordHashing.java [hunk 1]
```diff
@@ -100,7 +100,27 @@ public boolean verifyPasswordWithHash(String password, String hash) {
...
+        // Identifiers $2b, $2x and $2y are not recognized JBcrypt which only recognizes $2a, but, the actual hashed
+        // password can be verified using JBcrypt, so we replace the identifier for verification
+        if (hash.startsWith("$2b") || hash.startsWith("$2x") || hash.startsWith("$2y")) {
+            // we replace the unsupported identifier with $2a
+            return "$2a" + hash.substring(3);
+        }
+        return hash;
+    }
+
+    public boolean isInputAValidPasswordHash(String hash) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/emailpassword/PasswordHashing.java:118]: change name for make it more clear.

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/emailpassword/PasswordHashing.java:118]: What if i forget to modify this function??

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/emailpassword/PasswordHashing.java [hunk 2]
```diff
@@ -100,7 +108,29 @@ public boolean verifyPasswordWithHash(String password, String hash) {
...
+            return "$2a" + hash.substring(3);
+        }
+        return hash;
+    }
+
+    public boolean doesSuperTokensSupportInputPasswordHashFormat(String hash) {
+        // argon2 hash looks like $argon2id$v=..$m=..,t=..,p=..$tgSmiYOCjQ0im5U6...
+        // bcrypt hash starts with the algorithm identifier which can be $2a$, $2y$, $2b$ or $2x$,
+        // the number of rounds, the salt and finally the hashed password value.
+        return (hash.startsWith("$argon2id") || hash.startsWith("$2a") || hash.startsWith("$2x")
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/emailpassword/PasswordHashing.java:132]: also argon2 + others..

File: src/test/java/io/supertokens/test/emailpassword/UserMigrationTest.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+        // with bcrypt
+        {
+            String email = "test@example.com";
+            String plainTextPassword = "testPass123";
+            String passwordHash = "$2a$10$GzEm3vKoAqnJCTWesRARCe/ovjt/07qjvcH9jbLUg44Fn77gMZkmm";
+
+            // migrate user with passwordHash
+            EmailPassword.ImportUserResponse importUserResponse = EmailPassword
+                    .importUserWithPasswordHashOrUpdatePasswordHashIfUserExists(process.main, email, passwordHash);
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/emailpassword/UserMigrationTest.java:68]:
```suggestion
                    .importUserWithPasswordHash(process.main, email, passwordHash);
```

File: src/test/java/io/supertokens/test/emailpassword/UserMigrationTest.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        // test that sign in works
+        UserInfo userInfo = EmailPassword.signIn(process.main, email, password);
+        assertEquals(userInfo.email, email);
+        assertEquals(userInfo.passwordHash, passwordHash);
+
+        process.kill();
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STOPPED));
+    }
+
+    // test bcrypt with different salt rounds
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/emailpassword/UserMigrationTest.java:178]: wrong comment

File: src/test/java/io/supertokens/test/emailpassword/UserMigrationTest.java [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+            // test that sign in works
+            UserInfo userInfo = EmailPassword.signIn(process.main, email, password);
+            assertEquals(userInfo.email, email);
+            assertEquals(userInfo.passwordHash, passwordHash);
+        }
+
+        process.kill();
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STOPPED));
+    }
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/emailpassword/UserMigrationTest.java:257]: need to also add for different argon2 types + different config for 2id

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/emailpassword/EmailPassword.java [hunk 6]
```diff
@@ -76,6 +89,50 @@ public static UserInfo signUp(Main main, @Nonnull String email, @Nonnull String
...
+    public static ImportUserResponse importUserWithPasswordHash(Main main, @Nonnull String email,
+            @Nonnull String passwordHash, @Nullable PasswordHashingAlgorithm hashingAlgorithm)
+            throws StorageQueryException, StorageTransactionLogicException, UnsupportedPasswordHashingFormatException {
+
+        if (hashingAlgorithm != null) {
+            passwordHash = PasswordHashing.getInstance(main).updatePasswordHashWithPrefixIfRequired(hashingAlgorithm,
+                    passwordHash);
+        } else if (!PasswordHashing.getInstance(main).doesSuperTokensSupportInputPasswordHashFormat(passwordHash)) {
+            throw new UnsupportedPasswordHashingFormatException("Unsupported password hashing format");
+        }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/emailpassword/EmailPassword.java:101]:
```suggestion
            passwordHash = PasswordHashing.getInstance(main).updatePasswordHashWithPrefixIfRequired(hashingAlgorithm,
                    passwordHash);
      if (!PasswordHashing.getInstance(main).doesSuperTokensSupportInputPasswordHashFormat(passwordHash)) {
            throw new UnsupportedPasswordHashingFormatException("Unsupported password hashing format");
        }
```

File: src/main/java/io/supertokens/emailpassword/PasswordHashing.java [hunk 7]
```diff
@@ -94,13 +109,70 @@ private <T> T withConcurrencyLimited(Func<T> func) {
...
+    }
+
+    private static boolean isInputHashInArgon2Format(String hash) {
+        // argon2 hash looks like $argon2id or $argon2d or $argon2i $v=..$m=..,t=..,p=..$tgSmiYOCjQ0im5U6...
+        return (hash.startsWith("$argon2id") || hash.startsWith("$argon2i") || hash.startsWith("$argon2d"));
+    }
+
+    public String updatePasswordHashWithPrefixIfRequired(PasswordHashingAlgorithm hashingAlgorithm, String passwordHash)
+            throws UnsupportedPasswordHashingFormatException {
+        if (hashingAlgorithm.equals(PasswordHashingAlgorithm.ARGON2)) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/emailpassword/PasswordHashing.java:163]: do null check here

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Comment by jscyo** [comment]
## FLOWS
When passing hashing algorithm

Bcrypt + valid passwordHash ->  success
Bcrypt + passwordHash that is not bcrypt format -> 400 Error: password hash not in bcrypt format

Argon2 + validPasswordHash -> success
Argon2 + passwordHash that is not argon2 format -> 400 Error: password hash not in argon2 format 

Scrypt + validPasswordHash(supertokens scrypt format | random string) -> success
Scrypt + invalidPasswordHash (other password hashes format) -> 400 Error: password hash not in scrypt format

Without passing hashing algorithm

random string -> Error: invalid password hashing format

Bcrypt format password hash -> success
Argon2 format password hash -> success

Scrypt format password hash(supertokens format) -> success

File: src/test/java/io/supertokens/test/emailpassword/api/ImportUserWithPasswordHashAPITest.java [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+            assertTrue(e.statusCode == 400 && e.getMessage()
+                    .equals("Http error. Status Code: 400. Message: Unsupported password hashing algorithm"));
+        }
+
+        // passing hashingAlgorithm as bcrypt with random password hash
+        try {
+            JsonObject requestBody = new JsonObject();
+            requestBody.addProperty("email", "test@example.com");
+            requestBody.addProperty("passwordHash", "invalidHash");
+            requestBody.addProperty("hashingAlgorithm", "bcrypt");
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/emailpassword/api/ImportUserWithPasswordHashAPITest.java:172]: test with capital and random BcRyPT also

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**